tools/tplgfw-sanity-check.py

```diff
@@ -0,0 +1,140 @@
+            "fw_binary_header" / self.Fw_binary_header,
+            "modules" / self.Module_entry[this.fw_binary_header.module_entries],
+        )
+
+    def parse_file(self, filepath) -> ListContainer:
+        with open(filepath, "rb") as fw :
+            return self.Ext_manifest.parse_stream(fw)
+
+    def parse_fw_header(self, filepath) -> ListContainer:
+        with open(filepath, "rb") as fw :
+            return self.Ext_header.parse_stream(fw)
+
+if __name__ == "__main__":
+
+    def parse_cmdline():
+        parser = argparse.ArgumentParser(add_help=True, formatter_class=argparse.RawTextHelpFormatter, description="""
+        A Python tool to match UUID in topology and FW, then find out whether all module in topoloy
+        can be supported by this FW. Currently it only supports IPC4 FW and IPC3 & IPC4 topology.
+        The output includes FW version and with an error message if they are not matched. And all the
+        unsupported modules will be printed out.
+
+        example: tplgfw-sanity-check.py -t sof-tgl-nocodec.tplg -f basefw.bin"""
+        )
+
+        parser.add_argument('--version', action='version', version='%(prog)s 1.0')
+        parser.add_argument('-t', type=str, nargs='+', required=True, help="""Load topology files for check """)
+        parser.add_argument('-f', type=str, nargs='+', required=True, help="""Load firmware files for check """)
+        parser.add_argument('-v', '--verbose', action="store_true", help="Show check state")
+
+        return parser.parse_args()
+
+if __name__=="__main__":
+    def main():
+        tplgFormat = tplgtool2.TplgBinaryFormat()
+        cmd_args = parse_cmdline()
+        tplg = tplgtool2.GroupedTplg(tplgFormat.parse_file(cmd_args.t[0]))
+
+        fw_parser = Fw_Parser()
+        fw_content = fw_parser.parse_file(cmd_args.f[0])
+        print("FW release version: {0}.{1}".format(fw_content.ext_header.version_major, fw_content.ext_header.version_minor))
+
+        if fw_content.ext_header.id == '$AE1':
+            print("IPC4 supported FW\n")
+        elif fw_content.ext_header.id == 'XMan':
+            print("IPC3 supported FW\n")
+        else:
+            print("Invalid FW")
+            return -1
+
+        tplg_widget_map = {}
+        tplg_uuid_set = set()
+        fw_uuid_set = set()
+        for widget in tplg.widget_list:
+            name = widget["widget"]["name"]
+            widget_uuid = tplg.get_widget_uuid(widget)
+            if widget_uuid == None:
+                continue
+
+            tplg_widget_map[widget_uuid] = name
+            tplg_uuid_set.add(widget_uuid)
+
+        for m in fw_content.modules:
+            fw_uuid_set.add(m.uuid)
+
+        diff_set = tplg_uuid_set.difference(fw_uuid_set)
+        for item in diff_set:
+            failed_uuid = uuid.UUID(bytes=item)
+            print("wdiget {0} uuid: {1} is not found ".format(tplg_widget_map[item], failed_uuid))
```

Comment:
This will be simpler:
```
fw_uuid_set = {m.uuid for m in fw_content.modules}
tplg_uuid_dict = {tplg.get_widget_uuid(w): w["widget"]["name"]) for w in tplg.widget_list if w["widget"]["name"] is not None}

uuid_diff_set = set(tplg_uuid_dict.keys()) - fw_uuid_set

for uuid in uuid_diff_set:
	print(f'Widget {tplg_uuid_dict[uuid]} with UUID: {uuid} is not found')
```

Reply:
Thanks very much ! it is much more in python style.